File: demo-kafka-streams/src/main/java/demo/kafka/streams/listener/StreamListener.java
//package demo.kafka.streams.listener;
//
//import com.jayway.jsonpath.JsonPath;
//import demo.common.model.KafkaTopic;
//import org.apache.catalina.util.CustomObjectInputStream;
//import org.apache.kafka.common.serialization.Serdes;
//import org.apache.kafka.streams.KeyValue;
//import org.apache.kafka.streams.StreamsBuilder;
//import org.apache.kafka.streams.kstream.Grouped;
//import org.apache.kafka.streams.kstream.KStream;
//import org.apache.kafka.streams.kstream.Produced;
//import org.apache.kafka.streams.kstream.TimeWindows;
//import org.springframework.context.annotation.Bean;
//import org.springframework.stereotype.Component;
//
//import java.time.Duration;
//import java.util.Arrays;
//import java.util.stream.Stream;
//
//@Component
//public class StreamListener {
//
//    private final String INPUT_TOPIC = KafkaTopic.Topic.CHECKOUT_COMPLETE_TOPIC.getTopic();
//    private final String OUTPUT_TOPIC = KafkaTopic.Topic.CHECKOUT_OUTPUT_TOPIC.getTopic();
//
//    @Bean
//    public KStream<String, String> jsonToKStream(StreamsBuilder builder) {
//        String[] test = {"1","2"};
//        Arrays.stream(test).map(i -> i + "aa");
//        KStream<String, String> inputStream = builder.stream(INPUT_TOPIC);
//        inputStream
//                .map ((key, value) -> new KeyValue<>(getProductId(value), getAmount(value)) )
//                .groupByKey(Grouped.with(Serdes.Long(), Serdes.Long()) )
//                .windowedBy(TimeWindows.ofSizeAndGrace(Duration.ofMinutes(1), Duration.ZERO))
//                .reduce(Long::sum)
//                .toStream((key, value) -> key.key())
//                .mapValues(StreamListener::sendingJson)
//                .selectKey((key, value) -> null)
//                .to(OUTPUT_TOPIC, Produced.with(null, Serdes.String()));
//
//        return inputStream;
//    }
//
//
//
//    public static Long getProductId(String productId) {
//        return JsonPath.parse(productId).read("$.productId", Long.class);
//    }
//
//    public static Long getAmount(String amount) {
//        return JsonPath.parse(amount).read("$.amount", Long.class);
//    }
//
//    public static String sendingJson(Long productId, Long amount) {
//        String jsonData = "{\"productId\":%d,\"windowedAmount\":%d,\"createdAt\":%d}";
//        return String.format(jsonData, productId, amount, System.currentTimeMillis());
//    }
//
//}

File: demo-kafka-streams/src/main/java/demo/kafka/streams/listener/streamListener.kt
package demo.kafka.streams.listener

import demo.common.model.KafkaTopic
import demo.common.util.jsonValue
import demo.kafka.streams.model.Stream
import org.apache.kafka.common.serialization.Serdes
import org.apache.kafka.streams.KeyValue
import org.apache.kafka.streams.StreamsBuilder
import org.apache.kafka.streams.kstream.Grouped
import org.apache.kafka.streams.kstream.KStream
import org.apache.kafka.streams.kstream.Produced
import org.apache.kafka.streams.kstream.TimeWindows
import org.springframework.context.annotation.Bean
import org.springframework.stereotype.Component
import java.time.Duration

@Component
class StreamListener {

    val INPUT_TOPIC = KafkaTopic.Topic.CHECKOUT_COMPLETE_TOPIC.topic
    val OUTPIC_TOPIC = KafkaTopic.Topic.CHECKOUT_OUTPUT_TOPIC.topic

    @Bean
    fun kafkaStream(builder: StreamsBuilder): KStream<String, String> {
        return builder.stream<String, String>(INPUT_TOPIC)
            .apply {
                 map { _, value -> KeyValue(value.jsonValue("productId").toLong(), value.jsonValue("amount").toLong()) }
                .groupByKey( Grouped.with(Serdes.Long(), Serdes.Long()) )
                .windowedBy (TimeWindows.ofSizeAndGrace(Duration.ofMinutes(1), Duration.ZERO))
                .reduce(Long::plus)
                .toStream{ key, _ -> key.key() }
                .mapValues(Stream::sendingJson)
                .selectKey{ _, _ -> null }
                .to(OUTPIC_TOPIC, Produced.with(null, Serdes.String()))
            }
        }


}

